src/gretel_trainer/benchmark/entrypoints.py

```diff
@@ -0,0 +1,129 @@
+from __future__ import annotations
+
+import logging
+import shutil
+from inspect import isclass
+from pathlib import Path
+from typing import Callable, Optional, Type, Union, cast
+
+import pandas as pd
+from gretel_client.config import get_session_config
+
+from gretel_trainer.benchmark.core import BenchmarkConfig, BenchmarkException, Dataset
+from gretel_trainer.benchmark.custom.datasets import CustomDataset
+from gretel_trainer.benchmark.custom.models import CustomModel
+from gretel_trainer.benchmark.gretel.datasets import GretelDataset
+from gretel_trainer.benchmark.gretel.models import GretelModel
+from gretel_trainer.benchmark.job_spec import JobSpec, model_name
+from gretel_trainer.benchmark.session import Session, is_gretel_model
+
+logger = logging.getLogger(__name__)
+
+DatasetTypes = Union[CustomDataset, GretelDataset]
+ModelTypes = Union[
+    CustomModel,
+    Type[CustomModel],
+    GretelModel,
+    Type[GretelModel],
+]
+
+
+def compare(
+    *,
+    datasets: list[DatasetTypes],
+    models: list[ModelTypes],
+    config: Optional[BenchmarkConfig] = None,
+) -> Session:
+    cross_product: list[tuple[DatasetTypes, ModelTypes]] = []
+    for dataset in datasets:
+        for model in models:
+            cross_product.append((dataset, model))
+
+    return _entrypoint(jobs=cross_product, config=config, validate=_validate_compare)
+
+
+def launch(
+    *,
+    jobs: list[tuple[DatasetTypes, ModelTypes]],
+    config: Optional[BenchmarkConfig] = None,
+) -> Session:
+    return _entrypoint(jobs=jobs, config=config)
+
+
+def _entrypoint(
+    *,
+    jobs: list[tuple[DatasetTypes, ModelTypes]],
+    config: Optional[BenchmarkConfig] = None,
+    validate: Optional[Callable[[list[JobSpec]], None]] = None,
+) -> Session:
+    config = config or BenchmarkConfig()
+
+    working_dir_cleanup = None
+    if not config.working_dir.exists():
+        # clean up working_dir if we fail to prepare jobs
+        working_dir_cleanup = lambda: shutil.rmtree(config.working_dir)
+
+    try:
+        config.working_dir.mkdir(exist_ok=True)
+        job_specs = [
+            JobSpec(
+                dataset=_standardize_dataset(dataset, config.working_dir),
+                model=_create_model(model),
+            )
+            for dataset, model in jobs
+        ]
+        if validate is not None:
+            validate(job_specs)
+
+        if any(is_gretel_model(job) for job in job_specs):
+            _verify_client_config()
+
+    except Exception as e:
+        if working_dir_cleanup is not None:
+            working_dir_cleanup()
```

Comment:
tiny style nit: you could initially set `working_dir_cleanup = lambda: pass` and then just always call it here without checking for `None`.

Reply:
Or, hm, I guess `lambda: pass` is invalid, but `lambda: None` works.

Reply:
Great suggestion!